src/parquet/util/crypto.h

```diff
@@ -0,0 +1,240 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+
+#ifndef PARQUET_UTIL_CRYPTO_H
+#define PARQUET_UTIL_CRYPTO_H
+
+#include <openssl/aes.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <string>
+#include <sstream>
+#include <iostream>
+#include "parquet/exception.h"
+#include "parquet/types.h" 
+
+namespace parquet {
+ 
+inline int encrypt(Encryption::type alg_id, const uint8_t *plaintext, int plaintext_len, 
+                   uint8_t *key, int key_len, uint8_t *aad, int aad_len, 
+                   uint8_t *ciphertext) {
+                   
+  if (Encryption::PARQUET_AES_GCM_V1 != alg_id) {
+    std::stringstream ss;
+    ss << "Crypto algorithm " << alg_id << " currently unsupported\n";
+    throw parquet::ParquetException(ss.str());
+  }
+  
+  if (16 != key_len) {
+    std::stringstream ss;
+    ss << "Key length " << key_len << " currently unsupported\n";
+    throw parquet::ParquetException(ss.str());
+  }
+    
+  EVP_CIPHER_CTX *ctx;
```

Comment:
Manage ctx with RAII; otherwise any exception after line 63 is going to leak memory.

You could just use a unique_ptr with a custom deleter e.g.

```
struct EVPContextDeleter {
    void operator()(EVP_CIPHER_CTX *ctx) {
        if(ctx) { // would need to initialize to null in case there's a throw before the init
           EVP_CIPHER_CTX_free(ctx);
        }
    }
};

... stuff ...

std::unique_ptr<EVP_CYPHER_CTX, EVPContextDeleter> ctx;
```